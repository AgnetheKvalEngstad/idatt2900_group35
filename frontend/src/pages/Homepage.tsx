import TopicCard from "../components/TopicCard";
import CookieIcon from "@mui/icons-material/Cookie";
import { Grid2 } from "@mui/material";
import { Link } from "react-router-dom";
import { useTopics } from "../hooks/useTopics";
import { checkDifficulty } from "../utils/utils";

/**
 * A React component that renders a homepage that displays a grid of topic cards.
 * The topics are grouped by difficulty level, and each card
 * links to a corresponding topic page.
 *
 * @returns A React component that displays a loading state, error state,
 * or a grid of topic cards grouped by difficulty.
 */
export default function Homepage() {
  const cardIcon = CookieIcon;
  const cardSize: string = "medium";

  const { topics: data, loading, error } = useTopics();

  const sortedTopics = {
    ingen: data.filter(
      (topic) => checkDifficulty(topic.skillLevel) === "ingen"
    ),
    litt: data.filter((topic) => checkDifficulty(topic.skillLevel) === "litt"),
    mye: data.filter((topic) => checkDifficulty(topic.skillLevel) === "mye"),
  };

  if (loading) {
    return <div>Loading...</div>;
  }
  if (error) {
    return <div>Error loading topics</div>;
  }

  return (
    <div className="flex justify-center items-center flex-grow py-12">
      <Grid2
        container
        spacing={2}
        justifyContent="center"
        alignItems="center"
        style={{ maxWidth: "1200px" }}
      >
        {Object.values(sortedTopics).map((group, index) => (
          <Grid2
            key={index}
            size={{ xs: 12, sm: 4, md: 4 }}
            className="flex flex-col items-center gap-12 p-4"
          >
            {group.map((topic) => (
              <Link
                key={topic.id}
                to={`/topic/?difficulty=${checkDifficulty(topic.skillLevel)}`}
              >
                <TopicCard
                  cardTitle={topic.title}
                  cardIcon={cardIcon}
                  difficulty={checkDifficulty(topic.skillLevel)}
                  size={cardSize}
                />
              </Link>
            ))}
          </Grid2>
        ))}
      </Grid2>
    </div>
  );
}
