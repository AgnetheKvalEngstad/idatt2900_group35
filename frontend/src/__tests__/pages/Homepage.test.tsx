import React from "react";
import { render, screen, waitFor } from "@testing-library/react";
import Homepage from "../../pages/Homepage";
import axiosInstance from "../../api/axios";
import { BrowserRouter } from "react-router-dom";
import { vi } from "vitest";
import "@testing-library/jest-dom";

import type { AxiosInstance } from "axios";

vi.mock("../../api/axios");

const mockedAxios = axiosInstance as jest.Mocked<AxiosInstance>;

const mockTopics = [
  { id: 1, title: "Intro to Cookies", skillLevel: "0", userId: 1 },
  { id: 2, title: "Advanced Cookies", skillLevel: "2", userId: 2 },
];

const renderWithRouter = (ui: React.ReactElement) => {
  return render(<BrowserRouter>{ui}</BrowserRouter>);
};

describe("Tests for Homepage component", () => {
  beforeEach(() => {
    vi.clearAllMocks();
  });

  it("renders loading initially", async () => {
    mockedAxios.get.mockResolvedValueOnce({ data: mockTopics });

    renderWithRouter(<Homepage />);
    expect(screen.getByText("Loading...")).toBeInTheDocument();

    await waitFor(() => {
      expect(screen.queryByText("Loading...")).not.toBeInTheDocument();
    });
  });

  it("renders topic cards on success", async () => {
    mockedAxios.get.mockResolvedValueOnce({ data: mockTopics });

    renderWithRouter(<Homepage />);

    await waitFor(() => {
      expect(screen.getByText("Intro to Cookies")).toBeInTheDocument();
      expect(screen.getByText("Advanced Cookies")).toBeInTheDocument();
    });
  });

  it("shows error on failure", async () => {
    mockedAxios.get.mockRejectedValueOnce(new Error("Network error"));

    renderWithRouter(<Homepage />);

    await waitFor(() => {
      expect(screen.getByText("Error loading topics")).toBeInTheDocument();
    });
  });
});
