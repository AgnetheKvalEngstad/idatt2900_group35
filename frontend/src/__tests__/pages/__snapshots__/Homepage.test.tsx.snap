// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Tests for Homepage component > should render Homepage 1`] = `
<div
  className="flex justify-center items-center flex-grow py-12"
>
  <ForwardRef(Grid)
    alignItems="center"
    container={true}
    justifyContent="center"
    spacing={2}
    style={
      {
        "maxWidth": "700px",
      }
    }
  >
    <ForwardRef(Grid)
      padding={3}
      size={
        {
          "md": 4,
          "sm": 4,
          "xs": 6,
        }
      }
    >
      <TopicCard
        cardIcon={
          {
            "$$typeof": Symbol(react.memo),
            "compare": null,
            "type": {
              "$$typeof": Symbol(react.forward_ref),
              "render": [Function],
            },
          }
        }
        cardTitle="Insert title here"
        difficulty="ingen"
      />
    </ForwardRef(Grid)>
    <ForwardRef(Grid)
      padding={3}
      size={
        {
          "md": 4,
          "sm": 4,
          "xs": 6,
        }
      }
    >
      <TopicCard
        cardIcon={
          {
            "$$typeof": Symbol(react.memo),
            "compare": null,
            "type": {
              "$$typeof": Symbol(react.forward_ref),
              "render": [Function],
            },
          }
        }
        cardTitle="Insert title here"
        difficulty="litt"
      />
    </ForwardRef(Grid)>
    <ForwardRef(Grid)
      padding={3}
      size={
        {
          "md": 4,
          "sm": 4,
          "xs": 6,
        }
      }
    >
      <TopicCard
        cardIcon={
          {
            "$$typeof": Symbol(react.memo),
            "compare": null,
            "type": {
              "$$typeof": Symbol(react.forward_ref),
              "render": [Function],
            },
          }
        }
        cardTitle="Insert title here"
        difficulty="mye"
      />
    </ForwardRef(Grid)>
    <ForwardRef(Grid)
      padding={3}
      size={
        {
          "md": 4,
          "sm": 4,
          "xs": 6,
        }
      }
    >
      <TopicCard
        cardIcon={
          {
            "$$typeof": Symbol(react.memo),
            "compare": null,
            "type": {
              "$$typeof": Symbol(react.forward_ref),
              "render": [Function],
            },
          }
        }
        cardTitle="Insert title here"
        difficulty="ingen"
      />
    </ForwardRef(Grid)>
    <ForwardRef(Grid)
      padding={3}
      size={
        {
          "md": 4,
          "sm": 4,
          "xs": 6,
        }
      }
    >
      <TopicCard
        cardIcon={
          {
            "$$typeof": Symbol(react.memo),
            "compare": null,
            "type": {
              "$$typeof": Symbol(react.forward_ref),
              "render": [Function],
            },
          }
        }
        cardTitle="Insert title here"
        difficulty="litt"
      />
    </ForwardRef(Grid)>
    <ForwardRef(Grid)
      padding={3}
      size={
        {
          "md": 4,
          "sm": 4,
          "xs": 6,
        }
      }
    >
      <TopicCard
        cardIcon={
          {
            "$$typeof": Symbol(react.memo),
            "compare": null,
            "type": {
              "$$typeof": Symbol(react.forward_ref),
              "render": [Function],
            },
          }
        }
        cardTitle="Insert title here"
        difficulty="mye"
      />
    </ForwardRef(Grid)>
  </ForwardRef(Grid)>
</div>
`;
